Refactor

javierggt opened the pull request:
## Description

This PR fixes a couple of issues in PR #16 

- cyclical dependencies between `kalman_perigee_mon.py` and `monitor_win_perigee.py` 
- a few parameters that need to by passed to several functions, and which anyway have default values that we might not change.

What I did:
- moved all data-fetching code in `kalman_perigee_mon.py` and `monitor_win_perigee.py` into `kalman_watch_data.py`
- the data-fetching parts of `kalman_perigee_mon.EventPerigee` were moved into a class `EventPerigee` in `kalman_watch_data.py`. The plotting-related parts were kept, but the class was renamed `EventPerigeeMon`.
- Created a `config` that holds a `conf` object to keep track of globals while still allowing for changes.
- Created a `paths` module with functions to get a couple of paths used.
- the functions in `monitor_win_perigee.py` that were named like global variables (like `PERIGEES_INDEX_TABLE_PATH`) were moved to the `paths` module or as function members of `EventPerigee` (and renamed in all lower case).
- add a `--rad-data` command-line argument to tell where to find STK rad data. By default it is expected in `$SKA/data/stk/radiation_data.fits`

Notes:

There are other oddly named functions in `low_kalman_mon.py` that perhaps could go into the options/paths.

## Interface impacts
This PR requires the following Ska data: `$SKA/data/stk/radiation_data.fits`

## Testing

### Unit tests
<!-- At least one of these must be checked if unit tests exist. DELETE the unchecked/untested options. -->
No unit tests

### Functional tests

Scripts were run in three different ways:

- emulating a single call, exactly as it is in the task schedule (except data-dir is different):
  ```
  kalman_watch_kalman_perigee_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/branch  \
      --make-html
  kalman_watch_monitor_win_perigee \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/branch
  kalman_watch_low_kalman_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/branch
  ```
  [The output is here](https://icxc.cfa.harvard.edu/aspect/test_review_outputs/kalman_watch/pr-16/branch). In this case, the low-Kalman plot is empty because there is no data from previous runs.
- Using `kalman_watch_monitor_win_perigee` to produce a png of the monitoring window plot:
  ```
  kalman_watch_monitor_win_perigee \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/branch \
      --out-file mon_win_kalman_drops_-45d_-1d.png
  ```
  output: [mon_win_kalman_drops_-45d_-1d.png](https://icxc.cfa.harvard.edu/aspect/test_review_outputs/kalman_watch/pr-16/branch/mon_win_kalman_drops_-45d_-1d.png)
- successive calls to the script, imitating what the task schedule does, on a weekly basis to produce a history of data. **Note that this is NOT exactly what is run in the crontab/task_schedule**. This is because, in the crontab, the time range is automatically set to dates relative to the time the script is run. Also, the input/output file depends on this date. In this test, I ran with a changing date (not relative) so the file would change each time. The commands run are like these:
  ```
  kalman_watch_kalman_perigee_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/yearly \
      --make-html \
      --stop <stop>
  kalman_watch_monitor_win_perigee \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/yearly \
      --stop <stop> \
      --start <start> \
      --out-file=kalman_plot.html
  kalman_watch_low_kalman_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/yearly \
      --stop <stop> \
      --in-file=kalman_plot.html
  ```
    [output](https://icxc.cfa.harvard.edu/aspect/test_review_outputs/kalman_watch/pr-16/yearly)
- A single call of the scripts, based on the latest flight data. Copied the `weekly.2024-09-08_0015` snapshot of the kalman_watch3 data (found in `data/mta4`) into my data directory and ran:
   ```
   kalman_watch_kalman_perigee_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/run-once-on-latest \
      --make-html
   kalman_watch_monitor_win_perigee \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/run-once-on-latest
   kalman_watch_low_kalman_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/run-once-on-latest
   ```
   [output](https://icxc.cfa.harvard.edu/aspect/test_review_outputs/kalman_watch/pr-16/run-once-on-latest)
- a single call using incomplete radiation data so not all observations should have predicted IR flag data:
  ```
  kalman_watch_kalman_perigee_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/no-rad-data  \
      --rad-data=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/rad_data_truncated.fits \
      --make-html
  kalman_watch_monitor_win_perigee \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/no-rad-data \
      --rad-data=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/rad_data_truncated.fits
  kalman_watch_low_kalman_mon \
      --data-dir=$SKA/www/ASPECT_ICXC/test_review_outputs/kalman_watch/pr-16/no-rad-data
  ```
  [output](https://icxc.cfa.harvard.edu/aspect/test_review_outputs/kalman_watch/pr-16/no-rad-data). In this case, the last two perigees should have no predicted IR flag data.

javierggt committed:
moved data-related functions around

javierggt committed:
add options and paths modules

javierggt committed:
add --rad-data command-line option

javierggt committed:
using the same IR model numbers as in the notebook

javierggt committed:
remove unnecessary import

javierggt committed:
handle the case where there is not monitoring windows

javierggt committed:
add some stdout to low_kalman_mon for a typical error when debugging

javierggt committed:
bugfix

javierggt committed:
in kalman_watch_monitor_win_perigee, set default output name using start/stop and allow saving to png

javierggt committed:
remove spurious f

javierggt committed:
add to gitignore

javierggt committed:
changes to figure style

javierggt committed:
fix in kalman_watch_data cache: create directory including parents if it does not exist

javierggt committed:
use a mon_win_kalman_drops_-45d_-1d.html as default in-file in low_kalman_mon.py, and fix docstring in monitor_win_perigee.py

javierggt committed:
use default in/out filename in task schedule

javierggt committed:
use in-file from command line options in low_kalman_mon.py

taldcroft commented on kalman_watch/options.py:
```diff
@@ -0,0 +1,55 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Watch Kalman star data during perigee passages."""
```
For this module it would be good to follow the pattern from astropy and [four other Ska packages](https://github.com/search?q=org%3Asot%20ConfigNamespace&type=code) in terms of naming and import style. In particular `conf` is the configuration variable and the module is `config.py`.

javierggt replied:
I ended up removing the use of astropy config, but keeping the config/conf pattern, since that can be changed and is not really that much more complicated that a plain module.

taldcroft commented on kalman_watch/options.py:
```diff
@@ -0,0 +1,55 @@
+
+
+class ConfigNamespace(_config_.ConfigNamespace):
+    rootname = 'ska_trending'
```
See https://github.com/sot/kadi/blob/72bba923b62a44f73873fbb02738e2c4e7c3beda/kadi/config.py#L11 and the two links there.

This indicates that the rootname (`kalman_watch` not `ska_trending`) should be in a custom `ConfigItem`. I'll confess that it has been a while and putting the rootname in `ConfigItem` seems wrong. However, I do remember spending some time working on this in `kadi` and it seemed to work back then. OTOH it's possible things changed in astropy in the meantime, so this is something for testing and discussion.

javierggt replied:
The reason why I use this name is because I was thinking things would slowly converge into a single package (like Jean's ska_trend), and I would rather have one single configuration file rather than a collection of files.

I can change it back to use `kalman_watch` if that is what you really want.

I would note that I did experiment creating config files and it worked. I did see a bug where the file was not created (like your issue). I would have to check again what were the circumstances, but the name was not the cause of the misbehavior.

taldcroft commented:
In terms of functional testing, can you run a test starting with current flight data and then running once? I don't see a functional test showing that everything works in that case.

taldcroft commented:
In the new Perigee Kalman plot page, there are now four panels instead of three, but placed in the same overall figure size. Please expand the figure size so that the size of the subplots is similar to current.
E.g. https://icxc.cfa.harvard.edu/aspect/test_review_outputs/kalman_watch/pr-16/yearly/perigees/2024/Aug-23/index.html

javierggt commented:
> In terms of functional testing, can you run a test starting with current flight data and then running once? I don't see a functional test showing that everything works in that case.

What do you mean "with current flight data"? My first functional test was to run it once.

Do you mean that it uses the cache in `$SKA/www/ASPECT/kalman_watch3` but without writing the output in `$SKA/www/ASPECT/kalman_watch3`

taldcroft commented:
> What do you mean "with current flight data"? My first functional test was to run it once.

I am referring to "the low-Kalman plot is empty because there is no data from previous runs". There should be a demonstration that when we promote this code and run it the first time, it will produce the expected output which matches current flight except for expected differences.

taldcroft commented:
> the time range is automatically set to dates relative to the time the script is run.

@jeanconn would point out that this is another potential use case for a global ENV var that can change the default `NOW` time. This would really need to be done in both `cxotime` and `chandra_time` to be effective. The case above where the outputs depend on the input date range is slightly unusual but does occur. Hmm.

jeanconn commented:
Right - I'm not passionate about making code and changes for that NOW idea, but it does seem timely (pardon the pun) while Javier is spearheading a trending revamp to consider if something like that would help us to make some unit tests for trending packages and make it easier to do regression runs in trending packages.

javierggt commented:
> I am referring to "the low-Kalman plot is empty because there is no data from previous runs". There should be a demonstration that when we promote this code and run it the first time, it will produce the expected output which matches current flight except for expected differences.

That is exactly the reason why I then ran the third functional test, which runs multiple times, and consequently the plots is filled. This last test calls multiple times, so the last call is equivalent to calling the scripts once on current data.

javierggt commented:
are you worried that in test 3 I used `--stop <stop>`? and that somehow will cause something to be different?

javierggt commented:
I can run it now on the same folder as test 3, but without the `--stop <stop>` argument, so it would be the same as weekly processing. Would that be enough?

taldcroft commented:
In general, these tools write data files and then read them back again later. My request is to verify that the refactored code has not accidentally changed a filename. In that case the refactored code would self-consistently run fine and produce expected output, but it would fail when trying to run the first time with the data in `$SKA/www/ASPECT/kalman_watch3`.

javierggt commented:
If I copy the data in `$SKA/www/ASPECT/kalman_watch3` into a new location and run once on that, would that be enough?

taldcroft commented:
>  If I copy the data in $SKA/www/ASPECT/kalman_watch3 into a new location and run once on that, would that be enough?

Yes, that was the idea. Thanks.

javierggt committed:
increase height of perigee kalman plot

javierggt committed:
renamed options.OPTIONS -> config.conf

javierggt committed:
rename rootname in kalman_watch.config.ConfigNamespace

javierggt committed:
fix some pylance warnings

javierggt committed:
stop using astropy config

javierggt committed:
fixup

javierggt committed:
removed duplicated code related to get_kalman_drops_npnt:
- added n_cache to config
- renamed EventPerigee.get_kalman_drops_npnt as a property called kalman_drops_npnt,
- renamed get_kalman_drops_nman as a property called kalman_drops_nman,
- fixed wrong property caching for predicted_kalman_drops
- remove duplicated code in _get_kalman_drops_nman
- refactor get_kalman_drops_nman so it has the same signature as get_kalman_drops_npnt
- removed _get_binned_drops_from_npnt and use get_binned_drops_from_npnt instead,

javierggt committed:
set argparse defaults from config

javierggt committed:
ruff

javierggt committed:
rename radiation data file

javierggt commented:
@taldcroft, I made some changes that I believe the points you raised yesterday:
- got rid of astropy config
- added the radiation file to the paths module
- removed some config variable that are not necessary
- fixed the calls to argparse so the default value comes from the config
- removed some duplicate code (get_kalman_drops_nman/npnt)

taldcroft reviewed (approved):
@javierggt - this looks fantastic now, thanks! I'm really excited for this to go live.

jeanconn commented:
I have a quick go back in that I think the 'options' module referenced in the description was cut.  (I'm just recommending a final edit of the description).

javierggt commented:
I updated the description.

In my previous comment, I forgot to mention that I renamed the STK radiation data FITS file to `$SKA/data/stk/radiation_data.fits`. That way, any new STK data can go there, and we can update the file without changing the code.

taldcroft commented:
> In my previous comment, I forgot to mention that I renamed the STK radiation data FITS file

I noticed that and silently gave a :100: on it.
